File: src/main/java/no/novari/flyt/history/validation/OnlyOneStatusFilterValidator.java
package no.novari.flyt.history.validation;

import no.novari.flyt.history.model.instance.InstanceFlowSummariesFilter;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;
import java.util.Objects;

public class OnlyOneStatusFilterValidator implements ConstraintValidator<OnlyOneStatusFilter, InstanceFlowSummariesFilter> {

    @Override
    public boolean isValid(InstanceFlowSummariesFilter instanceFlowSummariesFilter, ConstraintValidatorContext constraintValidatorContext) {
        if (Objects.isNull(instanceFlowSummariesFilter)) {
            return true;
        }
        return !(Objects.nonNull(instanceFlowSummariesFilter.getStatuses())
                 && Objects.nonNull(instanceFlowSummariesFilter.getLatestStatusEvents()));
    }

}

File: src/main/kotlin/no/novari/flyt/history/validation/OnlyOneStatusFilterValidator.kt
package no.novari.flyt.history.validation

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext
import no.novari.flyt.history.model.instance.InstanceFlowSummariesFilter

class OnlyOneStatusFilterValidator : ConstraintValidator<OnlyOneStatusFilter, InstanceFlowSummariesFilter> {
    override fun isValid(
        instanceFlowSummariesFilter: InstanceFlowSummariesFilter?,
        context: ConstraintValidatorContext?,
    ): Boolean {
        if (instanceFlowSummariesFilter == null) {
            return true
        }

        return instanceFlowSummariesFilter.statuses == null || instanceFlowSummariesFilter.latestStatusEvents == null
    }
}
